pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,454 @@
+    int maxNumRecordsPerTask = mergeConfig.get(MergeRollupTask.MAX_NUM_RECORDS_PER_TASK_KEY) != null ? Integer.parseInt(
+        mergeConfig.get(MergeRollupTask.MAX_NUM_RECORDS_PER_TASK_KEY)) : DEFAULT_MAX_NUM_RECORDS_PER_TASK;
+    int numRecordsPerTask = 0;
+    List<List<String>> segmentNamesList = new ArrayList<>();
+    List<List<String>> downloadURLsList = new ArrayList<>();
+    List<String> segmentNames = new ArrayList<>();
+    List<String> downloadURLs = new ArrayList<>();
+
+    for (int i = 0; i < selectedSegments.size(); i++) {
+      OfflineSegmentZKMetadata targetSegment = selectedSegments.get(i);
+      segmentNames.add(targetSegment.getSegmentName());
+      downloadURLs.add(targetSegment.getDownloadUrl());
+      numRecordsPerTask += targetSegment.getTotalDocs();
+      if (numRecordsPerTask >= maxNumRecordsPerTask || i == selectedSegments.size() - 1) {
+        segmentNamesList.add(segmentNames);
+        downloadURLsList.add(downloadURLs);
+        numRecordsPerTask = 0;
+        segmentNames = new ArrayList<>();
+        downloadURLs = new ArrayList<>();
+      }
+    }
+
+    List<PinotTaskConfig> pinotTaskConfigs = new ArrayList<>();
+    for (int i = 0; i < segmentNamesList.size(); i++) {
+      Map<String, String> configs = new HashMap<>();
+      configs.put(MinionConstants.TABLE_NAME_KEY, offlineTableName);
+      configs.put(MinionConstants.SEGMENT_NAME_KEY,
+          StringUtils.join(segmentNamesList.get(i), MinionConstants.SEGMENT_NAME_SEPARATOR));
+      configs.put(MinionConstants.DOWNLOAD_URL_KEY,
+          StringUtils.join(downloadURLsList.get(i), MinionConstants.URL_SEPARATOR));
+      configs.put(MinionConstants.UPLOAD_URL_KEY, _clusterInfoAccessor.getVipUrl() + "/segments");
+
+      for (Map.Entry<String, String> taskConfig : taskConfigs.entrySet()) {
+        if (taskConfig.getKey().endsWith(MinionConstants.MergeRollupTask.AGGREGATION_TYPE_KEY_SUFFIX)) {
+          configs.put(taskConfig.getKey(), taskConfig.getValue());
+        }
+      }
+
+      configs.put(MergeRollupTask.MERGE_TYPE_KEY, mergeConfig.get(MinionConstants.MergeTask.MERGE_TYPE_KEY));
+      configs.put(MinionConstants.MergeTask.BUFFER_TIME_PERIOD_KEY,
+          mergeConfig.get(MinionConstants.MergeTask.BUFFER_TIME_PERIOD_KEY));
+      configs.put(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY,
+          mergeConfig.get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY));
+      configs.put(MinionConstants.MergeTask.ROUND_BUCKET_TIME_PERIOD_KEY,
+          mergeConfig.get(MinionConstants.MergeTask.ROUND_BUCKET_TIME_PERIOD_KEY));
+      configs.put(MinionConstants.MergeTask.MAX_NUM_RECORDS_PER_SEGMENT_KEY,
+          mergeConfig.get(MinionConstants.MergeTask.MAX_NUM_RECORDS_PER_SEGMENT_KEY));
+
+      configs.put(MergeRollupTask.SEGMENT_NAME_PREFIX_KEY,
+          MergeRollupTask.MERGED_SEGMENT_NAME_PREFIX + "_" + System.currentTimeMillis());
```

Comment:
So, are we providing a timestamp as a prefix?

```
merged_123451312321312_table-name_start_end_seq
```
is the expected segment name?

Reply:
Yes

Reply:
You might also want to include the merge level in the prefix?

Reply:
Added merge level for debugging purposes.

Reply:
We may have segment name collision if we schedule multiple tasks per round. Added sub-task number to handle this case.